src/pudl/settings.py

```diff
@@ -49,9 +54,65 @@ class GenericDatasetSettings(BaseModel):
     """
 
     disabled: bool = False
+    data_source: ClassVar[DataSource]
+
+    # TODO[pydantic]: This validator fails because it doesn't reproduce the current
+    # behavior we have on unspecified datasets.
+    # @model_validator(mode="before")
+    @classmethod
+    def validate_partitions_before(cls, data: dict[str, Any]) -> dict[str, Any]:
+        """Refactor with model validator after."""
+        for (
+            partition_name,
+            working_partitions,
+        ) in cls.data_source.working_partitions.items():
+            try:
+                partitions = data[partition_name]
+            except KeyError:
+                raise ValueError(
+                    f"{cls.__name__} is missing required '{partition_name}' field."
+                )
+
+            # If partition is empty or None, default to using all working_partitions
+            if not partitions:
+                data[partition_name] = working_partitions
+            else:
+                data[partition_name] = sorted(set(partitions))
+
+            nonworking_partitions = list(set(partitions) - set(working_partitions))
+            if nonworking_partitions:
+                raise ValueError(
+                    f"'{nonworking_partitions}' {partition_name} are not available."
+                )
 
-    @root_validator
-    def validate_partitions(cls, partitions):  # noqa: N805
+        return data
+
+    # TODO[pydantic]: This validator fails because the model is immutable after it's
+    # been defined.
+    # @model_validator(mode="after")
+    def validate_partitions_after(self: Self):
+        """Refactor with model validator after."""
+        for name, working_partitions in self.data_source.working_partitions.items():
+            try:
+                partition = getattr(self, name)
+            except KeyError:
+                raise ValueError(f"{self.__name__} is missing required '{name}' field.")
+
+            # If partition is None, default to working_partitions
+            if getattr(self, name) is None:
+                setattr(self, name, working_partitions)
+
+            nonworking_partitions = list(set(partition) - set(working_partitions))
+            if nonworking_partitions:
+                raise ValueError(f"'{nonworking_partitions}' {name} are not available.")
+            setattr(self, name, sorted(set(partition)))
+        return self
```

Comment:
Do we actually need to be mutating the object in this validator? We call `setattr` for two reasons:
1. If `partition` is `None. This will never be true because the default partition values will be used right?
2. Sorting the partitions. I don't think we have code that expects the partitions to be sorted.

Just my first thoughts, still need to play around with the code.

Reply:
I guess you're right. If every child class sets the default partition values based on the working partitions, then this is superfluous. Let me try making this change...

Reply:
It turns out this is currently how we're deduplicating input partitions.

Also, it woule be nice if we didn't have to reference `data_source.working_partitions` both when setting the data source and the defaults for each partition, since just having the data source should be enough to figure out what the default values are.

If we did it with `mode="before"` then when we get `None` for the partitions, we should be able to just fill it in with the working partitions from the class attribute.

Reply:
If we want to keep the values sorted we could programmatically create getters for each partition attribute that return sorted values?

Reply:
I agree I'd be nice if we didn't have to reference `data_source.working_partitions` both when setting the data source and the defaults for each partition. We could probably create the partition attributes from the data source information too though might not be worth the complexity. Something to think about if we want to do broader rewrite of our settings validation.

Instead of allowing duplicate partitions we could throw a validation error instead?

Reply:
Okay, now this is just raising errors if anything is wrong. Hopefully we don't have any waiting for us out there in the real world!